keps/487-kubectl-plugin/README.md

```diff
@@ -0,0 +1,572 @@
+
+```bash
+# kuctl workloads <name>
+kuctl workloads taco-*
+```
+
+| Name | Namespace | Command | Pods | Time | State |
+|------|-----------|---------|------|------|-------|
+| taco-345 | default | python | 3 | 2.322s | Runningg |
+| taco-123 | default | python | 2 |  | Pending |
+
+We likely want to also filter by state (or other attributes, TBA which others?)
+
+```bash
+kuctl workloads --state Pending
+```
+
+| Name | Namespace | Command | Pods | Time | State |
+|------|-----------|---------|------|------|-------|
+| taco-123 | default | python | 2 |  | Pending |
+
+
+### Describe Workloads
+
+Describe is intended to show more detailed information about one or more workloads. Akin to kubectl describe, we would stack them on top of the other. Unlike kubectl, I think we should have the -o json/yaml options here (it never made sense to me that kubectl uses describe for more rich metadata, but those output variables are available with "get" !
+
+```bash
+kuctl describe taco-123
+kuctl describe taco-123 taco-345
+```
+```console
+Feature Set A:
+  Field A            Field B     Field C
+  --------           --------    ------
+  cpu                650m (8%)   0 (0%)
+  memory             100Mi (0%)  0 (0%)
+Events:
+  Field A    Field B              Field C    
+  ----       -------              ------- ... 
+  Normal     Starting             Pancakes.
+```
+
+Note that I likely will develop this when I dig into working on the tool itself, and get a sense of all the attributes available to see about workloads. Right now I'm providing a generic template anticipating that. The above will include all metadata that the workload offers, and the additional features requested in the original prompt for reasons for pending or misconfiguration.
+
+The above should also provide different output formats:
+
+```bash
+kuctl describe taco-123 -o json
+kuctl describe taco-123 -o yaml
+```
+
+### Cancel Workloads
+
+A request to cancel would be akin to deleting the CRD. A "cancel" is more intuitive / natural than a delete request for this use case.
+This implementation will be tricky because we need to make the request to the underlying controller.
+
+
+```bash
+kuctl cancel taco-123
+```
+
+It might also be useful to request a cancel all, limited to the permission that the user has, a namespace, or other filter. This likely needs a confirmation.
+
+```bash
+kuctl cancel --all
+> Are you sure you want to cancel all workloads y/n?
+```
+
+Or without the prompt:
+
+```bash
+kuctl cancel --all --force
+```
+
+Or within a specific filter:
+
+```bash
+kuctl cancel --all --namespace insects
+```
+
+To support the multi-tenancy use case, filters for each of local-queue, cluster-queue and cohort will be allowed.
+
+### Watch Workload
+
+After submitting a workload, it's nice to be able to watch / stream logs. That should be easy to do.
```

Comment:
This may be require some thought on the CRD side.  

If a Job has multiple pods, I guess you could add an option to stream those specific logs.  

Events seem to make sense to me and then a separate call for logs where you allow for someone to drill into the particular pod.  

In cases like MPIJob or JobSet, you could have loads of different pods with different jobs so not sure what that interface would look like.  Maybe its post MVP?

Reply:
At least for the flux operator, we only care about the logs for one pod. I think for the MPI operator we'd care about a launcher pod? And JobSet is a total wild card! For those that are predictable, it might be possible to select the specific pod based on a label or similar. For others, we might need to step back and think of a larger strategy for handing logging (generally). It's definitely very important.

Reply:
I think that someone would want to see logs of a worker pod in MPI if they were debugging a pod.  

I also see job could be tricky (for same reason as JobSet) as you could have thousands of replica pods.  With indexed job as a label that could work but still requires some clarification on API.   

Maybe a kueue reviewer can weigh in on if logs should be included as a first round of the plugin.  I just don't see it as a trivial implementation (or even the API) and I think it could warrant discussion as a separate task.

Reply:
`kubectl logs job/{name}` actually gives the output for every pod in the job (with a prefix for the pod name).

In general, I think it will depend on the application and the user. Maybe we will have to have a command line flag to specify a name, index, or something else. But maybe default should be "everything"?

I think I'll defer here to the people that have more experience being end-users.
As an infrastructure developer, I'm biased to say that this is the application's problem, given how generic Kueue is.

Reply:
I can probably give better feedback about this when I start working on something.